powertool: change-state-into-spec can remove fields set by state-into-spec

justinsb opened the pull request:
- powertool: change-state-into-spec can remove fields set by state-into-spec
- tests: start to refactor normalization logic out of specific tests
- tests: Add ability to run CLI to script_test
- tests: Create change-state-into-spec powertool test scenario

justinsb committed:
powertool: change-state-into-spec can remove fields set by state-into-spec

Using the field manager, we can detect and remove any fields that were
set by state-into-spec.

justinsb committed:
tests: Create change-state-into-spec powertool test scenario

justinsb committed:
tests: start to refactor normalization logic out of specific tests

This way our various tests can share one normalization function

justinsb committed:
tests: Add ability to run CLI to script_test

This enables us to test our powertools in our existing framework

justinsb committed:
tests: Create test scenario for force-set-field powertool

Changing the location of a StorageBucket is a good example usage.

cheftako reviewed (dismissed):
Looks good.

maqiuyujoyce commented on pkg/stateintospec/remove.go:
```diff
@@ -0,0 +1,161 @@
+				managedFieldEntry.FieldsType)
+		}
+		fieldsV1 := managedFieldEntry.FieldsV1
+		if managedFieldEntry.FieldsV1 == nil {
```
Nit: `if fieldsV1 == nil`?

justinsb replied:
Good call, done

maqiuyujoyce reviewed (commented):
I went through the CLI change and the new stateintospec package: both look good to me except for some non-blocking questions/nits! Thank you so much for the clean code - it's very enjoyable to review it!

Do you think it's reasonable to split the test changes into a different PR? I'll leave it to you to decide.

justinsb commented on pkg/cli/powertools/changestateintospec/cmd.go:
```diff
@@ -0,0 +1,159 @@
+
+	o.FieldOwner = "change-state-into-spec"
+	o.NewStateIntoSpecAnnotation = "absent"
+	o.DryRun = false
```
We decided no, because long-term we want the default to be dry-run=false, and if we started with default is dry-run=true it would be hard to go to dry-run=false.

justinsb commented on pkg/cli/powertools/changestateintospec/cmd.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+func Run(ctx context.Context, stdout io.Writer, stderr io.Writer, options Options) error {
+	// log := klog.FromContext(ctx)
```
I believe so, we just aren't logging anything yet here

justinsb commented on pkg/cli/powertools/changestateintospec/cmd.go:
```diff
@@ -0,0 +1,159 @@
+	if options.ImpersonateUser == "" {
+		// Impersonate the KCC service account, which is allowed to make changes
+		options.ClusterOptions.Impersonate = &rest.ImpersonationConfig{
+			UserName: "system:serviceaccount:cnrm-system:cnrm-controller-manager-" + options.Namespace,
```
Yes, we should figure out a nice way to support that also!  This code is currently duplicated with our other powertool, would be good to refactor it out and have it e.g. read a flag or read the CCC

justinsb commented on pkg/cli/powertools/changestateintospec/cmd.go:
```diff
@@ -0,0 +1,159 @@
+
+	cmd.Flags().StringVar(&options.NewStateIntoSpecAnnotation, "set", options.NewStateIntoSpecAnnotation, "New value for the state-into-spec annotation")
+	cmd.Flags().BoolVar(&options.DryRun, "dry-run", options.DryRun, "dry-run mode will not make changes, but only print the changes it would make")
+	cmd.Flags().StringSliceVar(&options.PreserveFields, "keep-field", options.PreserveFields, "Additional fields to preserve in the spec")
```
I'll add more docs in another PR

justinsb replied:
Actually added examples and validation in this PR - I hadn't realized it was failing tests anyway

justinsb commented on pkg/cli/powertools/changestateintospec/cmd.go:
```diff
@@ -0,0 +1,159 @@
+
+	preserveFields := make(map[string]bool)
+	// The exception to state-into-spec; we always want to preserve the resourceID field.
+	preserveFields[".spec.resourceID"] = true
```
Ah ... we should incorporate this logic!

justinsb commented on pkg/cli/powertools/changestateintospec/cmd.go:
```diff
@@ -0,0 +1,159 @@
+	originalObject := u.DeepCopy()
+
+	preserveFields := make(map[string]bool)
+	// The exception to state-into-spec; we always want to preserve the resourceID field.
```
I think long term the generated ID goes into something under status.  "Spec in input, Status is output" or whatever our catchphrase is going to be :-)

cheftako replied:
+1. Beyond the philosophy point Justin brought up (which I agree with and I think is important). Separating out writing of the resource ID from the customer's, has 2 additional advantages. 1) It lets us know if we created or adopted the resource. 2) It allows us to understand if the resource is currently under KCCs control or not. (A bit racy but essentially KCC can only be controlling the resource once it has written the resource ID).

justinsb commented on pkg/cli/powertools/forcesetfield/cmd.go:
```diff
@@ -58,24 +58,15 @@ func AddCommand(parent *cobra.Command) {
 		RunE: func(cmd *cobra.Command, args []string) error {
 			ctx := cmd.Context()
 
-			if len(args) >= 1 {
```
Thanks - done!

justinsb commented on pkg/stateintospec/remove.go:
```diff
@@ -0,0 +1,161 @@
+		if element.FieldName != nil {
+			v, found := pos[*element.FieldName]
+			if !found {
+				return nil
```
Probably :-)  The intent is that this logic removes any field that we _know_ was set by state-into-spec=merge.  Maybe with the atomic lists change, the StorageBucket issue (in the tests) is not a problem

justinsb commented:
Added the missing headers and cleaned up some of the things here.

@maqiuyujoyce do you want to correct the field removal to work the same way as KCC?  I think to ease the direct actuation transition, we might want parity with the existing TF controllers for objects with state-into-spec=merge; specifically, we don't apply fields to GCP that are "owned by state-into-spec=merge" today.  At least that's my understanding.  So if we could easily get that behaviour (and share code with this powertool) that would be a great outcome.

maqiuyujoyce commented:
> @maqiuyujoyce do you want to correct the field removal to work the same way as KCC?

It'll be great if it does! Though we can do it incrementally. It should not be too hard to leverage the logic in https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/k8s/managedfields.go.

> I think to ease the direct actuation transition, we might want parity with the existing TF controllers for objects with state-into-spec=merge; specifically, we don't apply fields to GCP that are "owned by state-into-spec=merge" today.

That's the ideal design, yes. How much work will it add to the current direct controller implementation?
I assume currently most logic in a direct controller is customized so we can't easily use a shared logic to implement it?  

> So if we could easily get that behaviour (and share code with this powertool) that would be a great outcome.

Yes, it'll be great if the code logic in a generic controller (powertool/TF controller) can be shared with the direct controllers. If we agreed to proceed with it, then we can follow up offline to see what's the most efficient way to achieve it in the direct controllers.

justinsb commented:
> How much work will it add to the current direct controller implementation?  I assume currently most logic in a direct controller is customized so we can't easily use a shared logic to implement it?

I think we can roughly just put it here:  https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/master/pkg/controller/direct/directbase/directbase_controller.go#L214

There will be some gotchas around the fact that we currently write back to state, so maybe just an extra copy will be needed.

cheftako commented on config/tests/samples/create/harness.go:
```diff
@@ -529,11 +529,20 @@ func (h *Harness) GetClient() client.Client {
 			gvk := resource.GroupVersionKind()
 
+			// Special fake types for testing
+			if gvk.Group == "" && gvk.Kind == "RunCLI" {
```
Not a blocker but on the remote chance that this ever leaked onto a real cluster I would prefer a more obvious kind (eg. RunTestCLI).

cheftako commented on pkg/cli/powertools/forcesetfield/cmd.go:
```diff
@@ -53,29 +53,20 @@ func AddCommand(parent *cobra.Command) {
-					setFields[k] = v
+			for i := 0; i < len(args); i++ {
+				tokens := strings.SplitN(args[i], "=", 2)
+				if len(tokens) < 2 {
```
I think you are also in trouble if you get more than 2 tokens? Or are you assuming that the key will never have a '=' so the second '=' must be part of the value?

cheftako commented on pkg/cli/powertools/kubecli/options.go:
```diff
@@ -20,15 +20,27 @@ import (
 func (o *ClusterOptions) AddFlags(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&o.Kubeconfig, "kubeconfig", o.Kubeconfig, "Path to the kubeconfig file to use for CLI requests.")
+	cmd.Flags().StringVar(&o.ImpersonateUser, "as", o.ImpersonateUser, "Username to impersonate for the operation. User could be a regular user or a service account in a namespace.")
+	cmd.Flags().StringSliceVar(&o.ImpersonateGroups, "as-group", o.ImpersonateGroups, "Group to impersonate for the operation, this flag can be repeated to specify multiple groups.")
```
Not a blocker but it feels like a slice variable's flag should also be plural.

cheftako commented:
/hold for Justin to review minor comments before merging.

cheftako commented:
/lgtm
/approve
Ok with merging but it would be good to determine whether we want to pick up API comments before merging.

maqiuyujoyce reviewed (commented):
/lgtm
/approve

/hold cancel

All the comments should be non-blocking and we'll improve the powertool incrementally.

maqiuyujoyce committed:
Merge branch 'master' into powertools

maqiuyujoyce commented:
I tried to resolve the conflict using the UI (commit [1224acc](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/1646/commits/1224accf4b1f1deed1dabdee64240e8b7922c3a8)) but failed. While I can add a commit via GitHub UI, reverting it is not a supported feature.

I don't want to override this PR, so will create a separate PR.

ldanielmadariaga commented:
This item has been open for >12 months. As part of scheduled repository maintenance, this Issue/PR will be closed. If this is still a critical ask, please re-open and work with your Google Cloud Account team to prioritize it with GCP KCC Product Management.
